### PR TITLE
Detect if $HOME resides on networked filesystem and use $TMP instead

### PR DESCRIPTION
**Pull request by atamazov**
Implements NFS detection feature. The feature is intended to improve performance when the library is deployed on clusters. Clusters normally map home directories on network shares, which dramatically affects performance. Details:
- If the library sees that the planned db location is on the network, then the db is placed in the standard temporary directory.
- Works only for _user_ databases including user binary cache
- In effect only when MIOpen is used in non-DEV (non-developer) mode (MIOPEN_BUILD_DEV=Off)
- By-products (improvements of ExpandUser()):
  - [performance][NFC] Optimized performance (`HOME` env var is read only once now)
  - [quality][NFC] Replaced `/tmp` with portable `temp_directory_path()` when home path is incorrect
  - [quality][NFC] Returned boost path type instead of string (less casts where `ExpandUser()` is used)

## Testing

Manually tested:
- In non-DEV mode with local home directory
- In non-DEV mode with the same directory but mounted to NFS share
- Both initial runs (no databases and subdirectories created in $HOME) and subsequent runs were tested.

Logs:
- [local-1.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636028/local-1.log.txt)
- [local-2.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636027/local-2.log.txt)
- [nfs-1.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636024/nfs-1.log.txt)
- [nfs-2.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636029/nfs-2.log.txt)
- [dev-1.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636026/dev-1.log.txt)
- [dev-2.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11636025/dev-2.log.txt)

### [Informative] How to quickly configure NFS (Ubuntu 20.04):
SERVER_MACHINE
```bash
sudo apt install nfs-kernel-server
sudo systemctl start nfs-kernel-server.service
mkdir /tmp/home-exported
chmod 777 /tmp/home-exported/
# Add this to /etc/export:
# /tmp/home-exported CLIENT_MACHINE(rw,sync,no_subtree_check,no_root_squash)
sudo exportfs -a
```
CLIENT_MACHINE
```bash
mkdir /tmp/nfs-home
# Now you can run tests with HOME=/tmp/nfs-home & collect logs
# Example test:
#   MIOPEN_FIND_MODE=1 MIOPEN_LOG_LEVEL=6 HOME=/tmp/nfs-home ./bin/MIOpenDriver conv -F 1 -V 0 -i 1 |& tee logfile
# Then, remove user databases
sudo apt install nfs-common
sudo mount SERVER_MACHINE:/tmp/home-exported /tmp/nfs-home
# Note: To remove mount, use same command with "umount"
# Now you can run tests with HOME=/tmp/nfs-home again.
```

### CODE REVIEW

**Comment by JehandadKhan** [comment]
In general the approach seems fine to me, lets add some logging to make sure that new behavior is logged

File: src/expanduser.cpp [hunk 1]
```diff
@@ -1,25 +1,203 @@
...
+    struct statfs stat;
+    const int rc = statfs(path.c_str(), &stat);
+    if(rc != 0)
+    {
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        MIOPEN_LOG_NQE("statfs('" << path.string() << "') rc = " << rc << ", '" << strerror(errno)
+                                  << "'");
+        return false;
+    }
+    MIOPEN_LOG_NQI("Filesystem type at '" << path.string() << "' is: 0x" << std::hex << stat.f_type
```
Comment (atamazov) [inline src/expanduser.cpp:172]: @JehandadKhan Some additional logging here. `FindRecordUnsafe` prints the name of the file.

### Example for find-db.

If db file resides on EXT4 FS:
```
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '/home/atamazov/nfs-home/.config/miopen/' is: 0xef53 'EXT2/3/4_SUPER_MAGIC'
...
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 3-32-32-3x3-32-30-30-100-0x0-1x1-1x1-0-NCHW-FP32-F in cache for file /home/atamazov/nfs-home/.config/miopen//gfx1030_36.HIP.2_20_0_d51f49756.ufdb.txt
```
NFS case (FindRecordUnsafe prints ".../tmp/..."):
```
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config/miopen/'
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config/miopen'
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config'
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '/home/atamazov/nfs-home' is: 0x6969 'NFS_SUPER_MAGIC'
...
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 3-32-32-3x3-32-30-30-100-0x0-1x1-1x1-0-NCHW-FP32-F in cache for file /tmp/gfx1030_36.HIP.2_20_0_d51f49756.ufdb.txt
```

**Comment by atamazov** [comment]
@JehandadKhan @junliume @DrizztDoUrden @dmikushin Almost done, tested. You're welcome for review. Leftovers (TBD tomorrow morning):
- renaming expanduser.cpp/hpp
- adding detailed description.

**Comment by atamazov** [comment]
Decided not to rename expanduser.cpp/.hpp for now. I propose a larger refactoring: moving more things, including `ExpandUser()`, `IsNetworkedFilesystem()`, `LoadFile()`, `WriteFile()`, `GetLibPath()` and possibly the `TmpDir` class, to the special module `filesystem_utils`. It deserves a dedicated PR.

**Comment by atamazov** [comment]
Description updated. Ready for review & merge.

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/enhancement
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_normal
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/value_middle
- Proposed reviewers:
  - @JehandadKhan
  - @junliume
  - @DrizztDoUrden 
  - @dmikushin

**Review by DrizztDoUrden — APPROVED** [review]

File: src/expanduser.cpp [hunk 2]
```diff
@@ -35,23 +35,23 @@ MIOPEN_DECLARE_ENV_VAR(HOME)
...
-        // in tensoflow unit tests run via bazel, $HOME is not set, so this can happen
-        // setting home_dir to the /tmp for now
-        home_dir = boost::filesystem::temp_directory_path().native();
+        return {p};
     }
+    // todo:
+    // need to figure out what is the correct thing to do here
+    // in tensoflow unit tests run via bazel, $HOME is not set, so this can happen
+    // setting home_dir to the /tmp for now
+    return {boost::filesystem::temp_directory_path().native()};
```
Comment (dmikushin) [inline src/expanduser.cpp]: `temp_directory_path()` without args may throw exception: An error shall be reported `if !exists(p) || !is_directory(p)`, where p is the path to be returned. This is highly unlikely to happen, but still it's better to use `temp_directory_path(system::error_code& ec)` variant to avoid exceptions.

File: src/expanduser.cpp [hunk 3]
```diff
@@ -35,23 +35,23 @@ MIOPEN_DECLARE_ENV_VAR(HOME)
 
 namespace miopen {
 
-std::string ExpandUser(const std::string& path)
+static std::string GetHomeDir()
 {
-    std::string home_dir;
     const char* const p = GetStringEnv(HOME{});
     if(!(p == nullptr || p == std::string("/") || p == std::string("")))
```
Comment (dmikushin) [inline src/expanduser.cpp:185]: Is this thing `p == std::string("/")` checking first symbol is '/'? This style is very cryptic. Instead of these non-portable checks, please check p == nullptr and use boost again to check if the path exists, is a directory and is writable.

**Review by dmikushin — CHANGES_REQUESTED** [review]

File: src/expanduser.cpp [hunk 4]
```diff
@@ -133,22 +133,47 @@ bool IsNetworked(unsigned long ft)
...
+    //
+    // Solution (A): Just create_directories if path doesn't exist. It looks like a hack: if the
+    // path is on NFS, then the library will not use it, but will create directories (which is not
+    // good). And this function should not have any side effects on the file system.
+    //
+    // Solution (B): Traverse the input path up to the first existing directory, then check that
+    // directory. Stop after some fixed number of iterations to protect against possible file system
+    // problems. Let's use this solution for now.
+    auto path = path_;
+    for(int i = 0; i < 32; ++i)
```
Comment (dmikushin) [inline src/expanduser.cpp:154]: 1. Does this 32 denote the number of attempts to connect to the network filesystem? Please promote it to a macro with a name denoting this.

2. If this performs repeated attempts, should not a delay be added between them?

File: src/expanduser.cpp (hunk 1 above)
Comment (dmikushin) [reply src/expanduser.cpp:172]: Just to note: our solution handles one particular type of a network file system - NFS. There are more network filesystems out there, for example [Lustre](https://en.wikipedia.org/wiki/Lustre_(file_system)) is popular, but rarely used for $HOME.

**Comment by dmikushin** [comment]
@atamazov , does our database contain an attribution to a hostname? I mean, could a host identify whether or not a database found on disk actually belongs to him? If our NFS detection misses network filesystem presence, we could use the recorded hostname as an extra check to ensure the database does not apparently belong to a different host.

File: src/expanduser.cpp (hunk 1 above)
Comment (atamazov) [reply src/expanduser.cpp:172]: @dmikushin No, it handles many filesystem types, including LUSTRE.

File: src/expanduser.cpp (hunk 2 above)
Comment (atamazov) [reply src/expanduser.cpp]: @dmikushin If `temp_directory_path()` fails, then execution environment has really bad issues, similar to cases when malloc fails. The library can't do anything except failing anyway. If you insist, then I can add error handling but that would end with MIOPEN_THROW.

File: src/expanduser.cpp (hunk 3 above)
Comment (atamazov) [reply src/expanduser.cpp:185]: This code is not related to this PR. Please feel free to open an issue or PR if you think it's worth it.

File: src/expanduser.cpp (hunk 4 above)
Comment (atamazov) [reply src/expanduser.cpp:154]: 1. The source code says:
```C++
// Stop after some fixed number of iterations to protect against possible file system problems.
```
and `32` is used immediately after that. So the intent should be clear for the reader. Symbolic representation of this literal as a macro won't provide additional clarity, I guess.

And we use this literal only once. Therefore for now I do not see enough reasons for adding a macro.

Will do if you insist.

2. I do not think so. What the code does it _not_ repeated attempts to do the _same_ action. The action is different each time.

**Review by dmikushin — APPROVED** [review]

**Comment by atamazov** [comment]
@dmikushin The database names are designed in such a way that sharing them is a normal scenario and does not break the correctness (for more information, see the source code of the library and tickets in the private repository). But there are performance issues when miopen is deployed on supercomputing clusters, when hundreds-thousands of nodes simultaneously access the same shared file at the same time.

>  If our NFS detection misses network filesystem presence

Then the ticket should be filed. As a workaround, environment variables may be used.

> we could use the recorded hostname 

Good idea. But there could be cases when the user really wants to share databases. After this PR, the library will move them onto /tmp, but this can be easily overridden by environment variables. If hostname is encoded into a filename, then sharing won't work.

In other words, right now it is not clear which solution would be optimal.

Thanks for reviewing.

**Comment by atamazov** [comment]
This is https://github.com/ROCmSoftwarePlatform/MIOpen/labels/TESTING_CI_PASSED I guess?

**Review by junliume — APPROVED** [review]

File: src/db_path.cpp.in [hunk 5]
```diff
@@ -81,20 +81,35 @@ std::string GetSystemDbPath()
...
+    /// let's assume that the library is running on a cluster system, where home directory is shared
+    /// between all the nodes. In order to prevent the slowdown due to concurrent file access, let's
+    /// use the local temporary directory.
+    /// WARNING: This makes all the library's user databases non-persistent. However, for now, we
+    /// assume that the performance hit associated with the usage of shared home dir is worse than
+    /// the deletion of user databases.
+    const auto udb_path = ExpandUser("${MIOPEN_USER_DB_PATH}");
+#if !MIOPEN_BUILD_DEV
+    if(IsNetworkedFilesystem(udb_path))
+        return boost::filesystem::temp_directory_path();
```
Comment (atamazov) [inline src/db_path.cpp.in:102]: Warning can be added before this `return`...

File: src/binary_cache.cpp [hunk 6]
```diff
@@ -62,16 +62,28 @@ static boost::filesystem::path ComputeSysCachePath()
...
+        const std::string cache_dir = MIOPEN_CACHE_DIR;
+        const std::string version   = std::to_string(MIOPEN_VERSION_MAJOR)       //
+                                    + "." + std::to_string(MIOPEN_VERSION_MINOR) //
+                                    + "." + std::to_string(MIOPEN_VERSION_PATCH) //
+                                    + "." + MIOPEN_STRINGIZE(MIOPEN_VERSION_TWEAK);
+        p = miopen::ExpandUser(cache_dir) / version;
+#if !MIOPEN_BUILD_DEV
+        /// \ref nfs-detection
+        if(IsNetworkedFilesystem(p))
+            p = boost::filesystem::temp_directory_path();
```
Comment (atamazov) [inline src/binary_cache.cpp:84]: ...and here.

File: src/expanduser.cpp [hunk 7]
```diff
@@ -1,25 +1,203 @@
...
+    // The problem is that statfs doesn't work in this case, and we need to determine the type of FS
+    // _before_ the databases are created.
+    //
+    // Solution (A): Just create_directories if path doesn't exist. It looks like a hack: if the
+    // path is on NFS, then the library will not use it, but will create directories (which is not
+    // good). And this function should not have any side effects on the file system.
+    //
+    // Solution (B): Traverse the input path up to the first existing directory, then check that
+    // directory. Stop after some fixed number of iterations to protect against possible file system
+    // problems. Let's use this solution for now.
```
Comment (atamazov) [inline src/expanduser.cpp:152]: ⚓
